scenarios/validation_receipts/src/main.rs

```diff
@@ -0,0 +1,134 @@
+use holochain_types::prelude::*;
+use std::time::{Duration, Instant};
+use trycp_wind_tunnel_runner::embed_conductor_config;
+use trycp_wind_tunnel_runner::prelude::*;
+
+embed_conductor_config!();
+
+#[derive(Debug, Default)]
+pub struct ScenarioValues {}
+
+impl UserValuesConstraint for ScenarioValues {}
+
+fn agent_setup(
+    ctx: &mut AgentContext<TryCPRunnerContext, TryCPAgentContext<ScenarioValues>>,
+) -> HookResult {
+    connect_trycp_client(ctx)?;
+    reset_trycp_remote(ctx)?;
+
+    let client = ctx.get().trycp_client();
+    let agent_name = ctx.agent_name().to_string();
+
+    ctx.runner_context()
+        .executor()
+        .execute_in_place(async move {
+            client
+                .configure_player(agent_name.clone(), conductor_config().to_string(), None)
+                .await?;
+
+            client.startup(agent_name.clone(), None).await?;
+
+            Ok(())
+        })?;
+
+    install_app(ctx, scenario_happ_path!("crud"), &"crud".to_string())?;
+    try_wait_for_min_peers(ctx, Duration::from_secs(120))?;
+
+    Ok(())
+}
+
+fn agent_behaviour(
+    ctx: &mut AgentContext<TryCPRunnerContext, TryCPAgentContext<ScenarioValues>>,
+) -> HookResult {
+    let reporter = ctx.runner_context().reporter();
+
+    let action_hash: ActionHash = call_zome(
+        ctx,
+        "crud",
+        "create_sample_entry",
+        "this is a test entry value",
+        Some(Duration::from_secs(80)),
+    )?;
+
+    let response: Option<Record> = call_zome(
+        ctx,
+        "crud",
+        "get_sample_entry",
+        action_hash.clone(),
+        Some(Duration::from_secs(80)),
+    )?;
+
+    assert!(response.is_some(), "Expected record to be found");
+
+    let start = Instant::now();
+
+    'outer: loop {
+        let response: Vec<ValidationReceiptSet> = call_zome(
+            ctx,
+            "crud",
+            "get_sample_entry_validation_receipts",
+            action_hash.clone(),
+            Some(Duration::from_secs(80)),
+        )?;
+
+        // only check for complete if we actually get data back
+        // before any receipts are received, this is an empty vec
+        if !response.is_empty() {
+            let mut all_complete = true;
+
+            'inner: for set in response.iter() {
+                if !set.receipts_complete {
```

Comment:
I wonder how hard we want to push the concurrency here. While we're waiting for receipts to come back, we pause adding load to the system. Another option would be to push out actions as fast as possible and each behavior would just see how many of its previous actions have now got their validation receipts and stop tracking those actions.

I think both scenarios are valid but the second is more similar to how an active agent will apply load. They won't wait for validation receipts, they'll just create actions as fast as it suits them and we want to know how well Holochain can keep up.

Maybe we should have both?

Reply:
done. Added example charts for both to the top-level description of this PR